### PR TITLE
Security Blasters - Ill be back edition.

### PR DESCRIPTION
**Pull request by Arrow768**
Revival of #9640 by @Crosarius

### CODE REVIEW

File: code/modules/modular_computers/file_system/programs/security/guntracker.dm [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	filename = "guntracker"
+	filedesc = "Firearm Control"
+	extended_desc = "Official NTsec program for the tracking and remote control of wireless-enabled firearms."
+	program_icon_state = "security"
+	color = LIGHT_COLOR_ORANGE
+	size = 8
+	requires_ntnet = TRUE
+	available_on_ntnet = TRUE
+	required_access_download = access_armory
+//	required_access_run = access_armory
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/security/guntracker.dm:11]: Commented code bad.
```suggestion
```

File: code/modules/modular_computers/file_system/programs/security/guntracker.dm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	usage_flags = PROGRAM_CONSOLE
+
+/datum/computer_file/program/guntracker/ui_interact(var/mob/user)
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/security/guntracker.dm:14]:
```suggestion
```

Duplicated proc definition.

File: code/modules/modular_computers/file_system/programs/security/guntracker.dm [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		ui.auto_update_content = TRUE
+		. = TRUE
+
+/datum/computer_file/program/guntracker/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
+	. = ..()
+	data = . || data || list()
+
+	var/turf/T = get_turf(computer.loc)
+	var/list/wireless_firing_pins_data = list()
+	for(var/obj/item/device/firing_pin/wireless/P in wireless_firing_pins)
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/security/guntracker.dm:37]: Don't think wireless_firing_pins will ever contain anything not of that type.

```suggestion
	for(var/pin in wireless_firing_pins)
		var/obj/item/device/firing_pin/wireless/P = pin
```

File: code/modules/modular_computers/file_system/programs/security/guntracker.dm [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		if(!P.gun)
+			continue
+		var/turf/Ts = get_turf(P)
+		if(ARE_Z_CONNECTED(T.z, Ts.z))
+			var/list/guntracker_info = list(
+				"gun_name" = capitalize_first_letters(P.gun.name),
+				"registered_info" = P.registered_user,
+				"ref" = "\ref[P]",
+				"automatic_state" = (P.lockstatus == WIRELESS_PIN_AUTOMATIC),
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/security/guntracker.dm:46]: Ideally this stuff would be a single var that contains a state equal to one of the WIRELESS_PIN_X defines.

File: code/modules/projectiles/guns/energy/blaster.dm [hunk 5]
```diff
@@ -116,3 +116,62 @@
...
+
+/obj/item/gun/energy/disruptorpistol/verb/select_frame() // Let's people choose the aesthetic model of their gun. No stat changes.
+	set name = "Select Model"
+	set category = "Object"
+	set desc = "Click to select the model of your gun."
+
+	var/mob/M = usr
+	var/user_reply
+
+	if(!M.mind)	return FALSE
```
Comment (mikomyazaki) [inline code/modules/projectiles/guns/energy/blaster.dm:158]:
```suggestion
	if(!M.mind)
		return FALSE
```

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

**Review by TehFlaminTaco — APPROVED** [review]
Everything looks good to me. Not sure why this PR failed last time.

**Comment by Crosarius** [comment]
Scope creep and burnout made the last PR fail. To be frank I would prefer it if you did not use my code, because the mental anguish from the burnout from that PR is what caused me to quit this community and I feel that if you want all those features then you people can code it yourselves. However, I won't stop you, because I consider the code itself also the work of Myazaki because of how amazing they are and how much they helped me, so I'll leave the final decision on whether to utilise my code to them. However, I do not want my sprites to be used for this. Make your own.

Peace, and leave me alone.

File: code/modules/projectiles/guns/energy/blaster.dm [hunk 6]
```diff
@@ -116,3 +116,63 @@
...
+		return FALSE
+
+	user_reply = input("Select your frame.") in list("sub-compact","service","magnum")
+	if(!use_check_and_message(usr))
+		if(user_reply == "sub-compact")
+			icon = 'icons/obj/guns/disruptorpistol/disruptorpistolc.dmi'
+		if(user_reply == "service")
+			icon = 'icons/obj/guns/disruptorpistol/disruptorpistols.dmi'
+		if(user_reply == "magnum")
+			icon = 'icons/obj/guns/disruptorpistol/disruptorpistolm.dmi'
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy/blaster.dm:168]: this should be a radial menu so people can actually see what the icon will be

File: code/modules/projectiles/guns/energy/blaster.dm [hunk 7]
```diff
@@ -116,3 +116,63 @@
...
+/obj/item/gun/energy/disruptorpistol/verb/select_frame() // Let's people choose the aesthetic model of their gun. No stat changes.
+	set name = "Select Model"
+	set category = "Object"
+	set desc = "Click to select the model of your gun."
+
+	var/mob/M = usr
+	var/user_reply
+
+	if(!M.mind)
+		return FALSE
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy/blaster.dm:159]: pointless check, we should just do `if(use_check_and_message(usr)) return`

File: code/modules/projectiles/guns/energy/blaster.dm [hunk 8]
```diff
@@ -116,3 +116,63 @@
...
+/obj/item/gun/energy/disruptorpistol/security
+	pin = /obj/item/device/firing_pin/wireless
+
+/obj/item/gun/energy/disruptorpistol/verb/select_frame() // Let's people choose the aesthetic model of their gun. No stat changes.
+	set name = "Select Model"
+	set category = "Object"
+	set desc = "Click to select the model of your gun."
+
+	var/mob/M = usr
+	var/user_reply
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy/blaster.dm:156]: usr is typecasted to mob by default

File: code/modules/projectiles/guns/energy/blaster.dm (hunk 6 above)
Comment (fernerr) [reply code/modules/projectiles/guns/energy/blaster.dm:168]: Having the icon be tweakable to such a degree just feels weird in the first place, when they'll have the same stats no matter what. Unless I'm missing something.

**Review by mikomyazaki — APPROVED** [review]

**Pull request merged**